File: BOJ/3055.cpp
#include<iostream>
#include<vector>
#include<queue>
using namespace std;

const string kaktus = "KAKTUS";
const char stone = 'X';
const char water = '*';
const char emp = '.';
const char endpoint = 'D';
int r,c;
char map[50][50];
bool visited[50][50];
pair<int,int> start;
vector<pair<int,int>> waters;

void input() {
    cin>>r>>c;
    for(int i=0; i<r; i++) {
        for(int j=0; j<c; j++) {
            cin>>map[i][j];
            if(map[i][j]=='S') {
                map[i][j]=emp;
                start = {i,j};
            } else if(map[i][j]==water) {
                waters.push_back({i,j});
            }
        }
    }
}

bool check_range(int x,int y) {
    return x>=0 && x<r && y>=0 && y<c;
}

void spread_water() {
    int length = waters.size();
    for(int i=0; i<length; i++) {
        pair<int,int > w = waters[i];
        for(int i=0; i<4; i++) {
            int nx = w.first + "2011"[i] - '1';
            int ny = w.second + "1120"[i] - '1';
            if(!check_range(nx,ny)) {
                continue;
            }
            if(map[nx][ny]==water || map[nx][ny]==stone || map[nx][ny]==endpoint) {
                continue;
            }
            map[nx][ny] = water;
            waters.push_back({nx,ny});
        }
    }
}

string solution() {
    queue<pair<pair<int,int>,int>> q;
    q.push({start,0});
    visited[start.first][start.second] = true;
    int before = -1;
    while(!q.empty()) {
        int fx = q.front().first.first;
        int fy = q.front().first.second;
        int count = q.front().second;
        q.pop();
        if(map[fx][fy]==endpoint) {
            return to_string(count);
        }
        if(before!=count) {
            spread_water();
            before = count;
        }
        ++count;
        for(int i=0; i<4; i++) {
            int nx = fx + "2011"[i] - '1';
            int ny = fy + "1120"[i] - '1';

            if(!check_range(nx,ny)) {
                continue;
            }
            if(visited[nx][ny]) {
                continue;
            }
            if(map[nx][ny]==water || map[nx][ny]==stone) {
                continue;
            }
            visited[nx][ny] = true;
            q.push({{nx,ny},count});
        }
    }
    return kaktus;
}

void solve() {
    input();
    cout<<solution();
}

int main() {
    cin.tie(NULL);
    cout.tie(NULL);
    ios_base::sync_with_stdio(false);

    solve();
}

File: JAVA/3055.java
import java.io.*;
import java.util.*;

public class JUN3055_이경민 {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;

    static class Pair<A,B> {
        A x;
        B y;
        Pair(A x,B y) {
            this.x = x;
            this.y = y;
        }
    }

    static final String kaktus = "KAKTUS";
    static final char stone = 'X';
    static final char water = '*';
    static final char empty = '.';
    static final char exit = 'D';

    static int r,c;
    static char[][] map;
    static boolean[][] visited;
    static Pair<Integer,Integer> start;
    static ArrayList<Pair<Integer,Integer>> waters = new ArrayList<>();

    public static void input() throws Exception {
        st = new StringTokenizer(br.readLine());
        r = Integer.parseInt(st.nextToken());
        c = Integer.parseInt(st.nextToken());
        visited = new boolean[r][c];
        map = new char[r][c];
        for(int i=0; i<r; i++) {
            String str = br.readLine();
            for(int j=0; j<c; j++) {
                map[i][j] = str.charAt(j);
                if(map[i][j]=='S') {
                    map[i][j] = empty;
                    start = new Pair(i,j);
                } else if(map[i][j]==water) {
                    waters.add(new Pair(i,j));
                }
            }
        }
    }

    public static boolean checkRange(int x,int y) {
        return x>=0 && y>=0 && x<r && y<c;
    }

    public static void spreadWater() {
        int size = waters.size();
        for(int k=0; k<size; k++) {
            Pair<Integer,Integer> w = waters.get(k);
            for(int i=0; i<4; i++) {
                int nx = w.x + "2011".charAt(i) - '1';
                int ny = w.y + "1120".charAt(i) - '1';
                if(!checkRange(nx,ny)) {
                    continue;
                }
                char ch = map[nx][ny];
                if(ch==water||ch==stone||ch==exit) {
                    continue;
                }
                map[nx][ny] = water;
                waters.add(new Pair(nx,ny));
            }
        }
    }

    public static String solve() {
        ArrayDeque<Pair<Pair<Integer,Integer>,Integer>> q = new ArrayDeque<>();
        visited[start.x][start.y] = true;
        q.add(new Pair(start,0));
        int before = -1;
        while(!q.isEmpty()) {
            int fx = q.getFirst().x.x;
            int fy = q.getFirst().x.y;
            int count = q.pollFirst().y;
            if(map[fx][fy]==exit) {
                return Integer.toString(count);
            }
            if(before!=count) {
                before=count;
                spreadWater();
            }
            ++count;
            for(int i=0; i<4; i++) {
                int nx = fx + "2011".charAt(i) - '1';
                int ny = fy + "1120".charAt(i) - '1';
                if(!checkRange(nx,ny)) {
                    continue;
                }
                if(visited[nx][ny]) {
                    continue;
                }
                char ch = map[nx][ny];
                if(ch==water || ch==stone) {
                    continue;
                }
                visited[nx][ny] = true;
                q.addLast(new Pair(new Pair(nx,ny),count));
            }
        }
        return kaktus;
    }

    public static void main(String[] args) throws Exception {
        input();
        System.out.print(solve());
    }
}
